llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp

```diff
@@ -883,9 +883,15 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .clampMaxNumElements(1, p0, 2);
 
   getActionDefinitionsBuilder(G_INSERT_VECTOR_ELT)
-      .legalIf(typeInSet(0, {v16s8, v8s8, v8s16, v4s16, v4s32, v2s32, v2s64}))
+      .legalIf(
+          typeInSet(0, {v16s8, v8s8, v8s16, v4s16, v4s32, v2s32, v2s64, v2p0}))
       .moreElementsToNextPow2(0)
```

Comment:
`.legalFor({{v16s8, s32, s64}, {v8s8, s32, s64}, {v8s16, s32, s64}, {v4s16, s32, s64}, {v4s32, s32, s64}, {v2s32, s32, s64}, {v2s64, s64, s64}, {v2p0, p0, s64}})`

Reply:
Those would not really be correct for fp types.

Reply:
```
// Generic insertelement.
def G_INSERT_VECTOR_ELT : GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type0:$src, type1:$elt, type2:$idx);
  let hasSideEffects = false;
}
```
We have to specify all three types in the legalizer: in/out vector, the element, and the index.

Reply:
Ah, the famous f16 again.

Reply:
We need floats in the LLT. `{v8i16, i16, i64}` is illegal. There are no i16s. `{v8f16, f16, i64}` is illegal because of the famous f16.

Reply:
There is no notion of fpr and gpr in the legalizer. We have to state the legal types for operations. The issue comes up with untyped operations, like e.g., G_SELECT, G_INSERT_VETOR_ELT, G_BUILD_VECTOR, ... 

For typed operations, we need no floats in the LLT. s16 is legal for G_FADD, but it is not legal for G_ADD. In the later case, it would have been an i16.

Reply:
> There is no notion of fpr and gpr in the legalizer. We have to state the legal types for operations. The issue comes up with untyped operations, like e.g., G_SELECT, G_INSERT_VETOR_ELT, G_BUILD_VECTOR, ...
> 
> For typed operations, we need no floats in the LLT. s16 is legal for G_FADD, but it is not legal for G_ADD. In the later case, it would have been an i16.

I'm not sure I am understanding your comment, but I think I agree with the first half.

For the second half the problem is not between G_FADD and G_ADD, it is between G_FADD with fp16 and G_FADD with bf16. They both have the same s16 sized operands, but there is currently no way to tell them apart in GISel.